<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>HW1</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h1 id="homework-1">Homework 1</h1>
<p>Homework can be submitted via the github link which will create a repository for you with basic template of files you can edit to solve the homework. See the the table for homework submission links which will help you create a github repository in the class team.cd <a href="https://piazza.com/ucr/fall2020/gen220/resources">https://piazza.com/ucr/fall2020/gen220/resources</a></p>
<ol type="1">
<li>Write a script called <code>download_count.sh</code> which does the following.
<ul>
<li>Download the data file <a href="https://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec_Core">https://ftp.ncbi.nlm.nih.gov/pub/UniVec/UniVec_Core</a> from NCBI</li>
<li>Print out the count of the number of FASTA format sequences in this file - see <a href="https://en.wikipedia.org/wiki/FASTA_format">Wikipedia FASTA format</a> - each record starts with a <code>&gt;</code></li>
</ul></li>
<li>Write a script called <code>summary_exons.sh</code> which summarizes the total length of exons in the file <a href="https://raw.githubusercontent.com/biodataprog/GEN220_data/main/data/rice_random_exons.bed">data/rice_random_exons.bed</a>. These data are in the BED file format. The columns are “Chromosome”, “Start position”, “Stop position”. The length of a feature (or exon in this case) is computed by doing the computation: STOP - START
<ul>
<li>read in the file</li>
<li>use a loop structure to read each line</li>
<li>add up the length of each exon by summing this into a variable</li>
<li>Print out the total length of exon features at the end.</li>
<li>You do not need to save this for each chromosome, just print out the total length for this example - however if this is too easy for you, go ahead and make a more sophisticated report which presents, per chromosome, the total length of exons as well as the total number of exons, and the average length of exons.</li>
</ul></li>
<li>Write a script called <code>strand_gene_count.sh</code> to calculate the number of genes that are on the positive (+) and negative (-) strand in the file. This file format is called <a href="https://en.wikipedia.org/wiki/General_feature_format">GFF</a> - the strand of the gene is encoded in 7th column.</li>
</ol>
<ul>
<li><a href="https://fungidb.org/common/downloads/release-48/ScerevisiaeS288c/gff/data/FungiDB-48_ScerevisiaeS288c.gff">https://fungidb.org/common/downloads/release-48/ScerevisiaeS288c/gff/data/FungiDB-48_ScerevisiaeS288c.gff</a></li>
</ul>
</body>
</html>
